docs/Guide/Syntactic-Macros.md

```diff
@@ -0,0 +1,102 @@
+Syntactic macros allow new language constructs to be created.  See [Macros](Macros) for a list of existing ones in boo.
+
+We could, for example, mimic VisualBasic's with statement.
+
+Given the following code
+
+```boo
+fooInstanceWithReallyLongName = Foo()
+fooInstanceWithReallyLongName.f1 = 100
+fooInstanceWithReallyLongName.f2 = "abc"
+fooInstanceWithReallyLongName.DoSomething()
+```
+
+If we define a 'with' macro we could rewrite it like this:
+
+```boo
+with fooInstanceWithReallyLongName:
+    _f1 = 100
+    _f2 = "abc"
+    _DoSomething()
+```
+
+In boo, macros are CLI objects that implement the Boo.Lang.Compiler.IAstMacro interface. It is interesting to note that there is nothing magic about these objects. They must simply implement the interface the compiler expects. This implies that boo macros can be written in any CLI language!
+
+When an unknown syntactic structure is encountered at compile time, like the with statement above, the compiler will look for the correct IAstMacro class, create an instance, and ask that instance to expand the macro. The compiler identifies the class to use via a simple naming convention. The class name must start with the name of the macro and end with 'Macro'. Additionally the [Pascal case](https://github.com/bamboo/boo/wiki/Pascal-case) naming convention must be used. So in this case the name must be 'WithMacro'.
+
+Here's the code to implement our macro:
+
+```boo
+
+import Boo.Lang.Compiler
+import Boo.Lang.Compiler.Ast
+import Boo.Lang.Compiler.Ast.Visitors
+ 
+class WithMacro(AbstractAstMacro):
```

Comment:
It might be nice to rewrite this in the idiomatic style as a MacroMacro invocation.